File: Core/YABACore/src/commonMain/kotlin/dev/subfly/yabacore/state/creation/annotation/AnnotationCreationEvent.kt
package dev.subfly.yabacore.state.creation.annotation

import dev.subfly.yabacore.model.annotation.ReadableSelectionDraft
import dev.subfly.yabacore.model.utils.YabaColor

sealed class AnnotationCreationEvent {
    /** Initialize for creating a new annotation from reader selection. */
    data class OnInitWithSelection(val draft: ReadableSelectionDraft) : AnnotationCreationEvent()

    /** Initialize for editing an existing annotation. Requires bookmarkId for loading. */
    data class OnInitWithAnnotation(
        val bookmarkId: String,
        val annotationId: String,
    ) : AnnotationCreationEvent()

    data class OnSelectNewColor(val newColor: YabaColor) : AnnotationCreationEvent()
    data class OnChangeNote(val note: String) : AnnotationCreationEvent()
    data class OnSave(
        val onSavedCallback: () -> Unit,
        val onErrorCallback: (Throwable) -> Unit,
    ) : AnnotationCreationEvent()

    data class OnDelete(
        val onDeletedCallback: () -> Unit,
        val onErrorCallback: (Throwable) -> Unit,
    ) : AnnotationCreationEvent()
}

File: Darwin/YABACore/State/Creation/Annotation/AnnotationCreationEvent.swift
//
//  AnnotationCreationEvent.swift
//  YABACore
//
//  Parity with Compose `AnnotationCreationEvent`.
//

import Foundation

public enum AnnotationCreationEvent: Sendable {
    case onInitWithSelection(ReadableSelectionDraft)
    case onInitWithAnnotation(bookmarkId: String, annotationId: String)
    case onSelectNewColor(YabaColor)
    case onChangeNote(String)
    case onChangeQuote(String?)
    case onSave
    case onDelete

    /// Direct insert (legacy one-shot).
    case create(
        bookmarkId: String,
        type: AnnotationType,
        note: String?,
        quote: String?
    )
}
